### DISCUSSION TITLE
trouble with pixel.energyCompensation blowing out the specular

### DISCUSSION THREAD

**Discussion by EnviralDesign** [google/filament#4989] [Q&A] [answered]
Greetings, I've been slowly but surely implementing the shader side of things in a TouchDesigner project. It involves a lot breaking and reconnecting things etc.

Right now I'm struggling to figure out what I've got connected in correctly that's causing the particular effect I'm seeing. Here's a sphere, lit only with IBL, and using the venetian_crossroads_2k file run through cmgen.

![image](https://user-images.githubusercontent.com/10091486/146798349-a1a037f2-ea5f-488e-aa54-2bdb888475e7.png)

Here's an "approximate" example of what I think it should look like (managed this by fudging pixel.dfg.xy = vec2(0.1) ):
![image](https://user-images.githubusercontent.com/10091486/146798504-837fb727-d71b-4fb2-bc5c-a0513905543b.png)

I think I've traced the issue down to this line in **evaluateIBL( )**
```Fr *= singleBounceAO(specularAO) * pixel.energyCompensation;```

Multiplying pixel.energyCompensation against Fr causes the values to sky rocket. Here's an example of a value in the middle of the sphere:
![image](https://user-images.githubusercontent.com/10091486/146800911-f7922d44-4a58-4411-a00f-e91a85a8abe2.png)

So I traced back to see where energyCompensation is getting set from, and came to **getEnergyCompensationPixelParams()**
This line specifically: ``` pixel.energyCompensation = 1.0 + pixel.f0 * (1.0 / pixel.dfg.y - 1.0); ```

I noticed that the specific part of that equation that is giving really high numbers is this part:
``` 1.0 / pixel.dfg.y ``` because pixel.dfg.y is made up of extremely small numbers. Here's what pixel.dfg.xy looks like mapped to r/g (no gamma or post stuffs)
![image](https://user-images.githubusercontent.com/10091486/146802221-fcbd4a06-7827-453d-9767-9416e478b4a4.png)

I thought maybe this was being calculated incorrectly, so I went to the place where pixel.dfg.xy is calculated, which is this function:
```
vec3 PrefilteredDFG_LUT(float lod, float NoV) {
    return textureLod(light_iblDFG, vec2(NoV, lod), 0.0).rgb;
```

The texture for me looks like the one on the documentation, and was generated by cmgen:
![image](https://user-images.githubusercontent.com/10091486/146802602-adefde44-f392-486f-942f-903779a0d6cc.png)

I mapped out the lod and NoV arguments to the sphere as well to see if anything looked funny, but so far as I could tell looked pretty standard:

NoV:
![image](https://user-images.githubusercontent.com/10091486/146802977-0fd60b96-d095-4b2d-af4e-53c843b4b85a.png)

lod (roughnessPerceptual I believe):
![image](https://user-images.githubusercontent.com/10091486/146803119-c48ed48d-fd73-4f3b-b4d9-55c237863fb5.png)

---

So I'm at the moment stumped. Any help would be much appreciated!

**Comment by romainguy** [comment, marked as answer]
cmgen has a flag to generate a LUT suitable for multi scattering and energy compensation.

**Reply by EnviralDesign** [reply]
@romainguy thank you for the quick response, I figured it was going to be a facepalm but I was also looking in the wrong place. Thank you so much. That did the trick.
